NXP S32 CANXL correct the calculating initial timing

congnguyenhuu opened the pull request:
Add a fuction that calculates the data phase bit timing so that the prescaler of the data phase bit timing must be equal this one
of the nominal bit timing. Due to the hardware CANXL use the same bit time quanta prescaler for the nominal bit timing and the data phase bit timing.
Fixes #77436

congnguyenhuu commented:
Hello @henrikbrixandersen, could you please visit this PR? Thanks.

congnguyenhuu commented:
Hello @henrikbrixandersen, could you please visit this PR? Thanks

henrikbrixandersen reviewed (changes requested):
Thanks. I do see the issue, but I do not agree with the proposed solution (we cannot just expose an `*_internal` function because you need it in one driver).

I will work on a solution that is more generic and suits a few other use-cases, I have in mind.

congnguyenhuu commented:
hello @henrikbrixandersen, have you got any more solution idea for this? what the release will you target to solve.
I would like to target to solve this in release 4.0.0
If you have any idea, you can share with me. I can work on it if possible.

manuargue commented:
> Thanks. I do see the issue, but I do not agree with the proposed solution (we cannot just expose an `*_internal` function because you need it in one driver).

I agree with this comment, this is not an acceptable way to solve the issue.

> hello @henrikbrixandersen, have you got any more solution idea for this? what the release will you target to solve.
I would like to target to solve this in release 4.0.0

@congnguyenhuu I'd suggest to either discuss with the CAN maintainers on a better solution and address the code changes on this pr; or to close this pr and log a feature request explaining what is the use-case need if you have any proposals of how to address it. I'm sure the CAN maintainers will address it at a given time, but 4.0.0 is round the corner.

congnguyenhuu committed:
drivers: can: introduce new calculate timing API with timming limits

Add 2 APIs can_calc_timing_with_limits() and
can_calc_timing_data_with_limits() with timing limits the minimum and
maximun allowed timing parameter values.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
drivers: nxp_s32_canxl: correct the calculating initial timing

Use API can_calc_timing_data_with_limits() to calculate the timing
with a limit to ensure that the prescaler for the data phase bit
timing matches that of the nominal bit timing. This is necessary
because the CANXL hardware uses the same bit time quanta prescaler
for both the nominal and data phase bit timings.

Update the maximum values of timing that is allowed by hardware.a

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu commented:
Hello @henrikbrixandersen, I have updated to introduce new calculate timing APIs  with timing limits that allow to pass parameters for the minimum and maximum allowed timing parameter values.

dleach02 commented:
> Thanks. I do see the issue, but I do not agree with the proposed solution (we cannot just expose an `*_internal` function because you need it in one driver).
> 
> I will work on a solution that is more generic and suits a few other use-cases, I have in mind.

@henrikbrixandersen have you been able to flush out a more generic solution?

henrikbrixandersen commented:
> @henrikbrixandersen have you been able to flush out a more generic solution?

Please see https://github.com/zephyrproject-rtos/zephyr/issues/77436#issuecomment-2606836985
